Implement "Filter by location"

mtiessen1175 opened the pull request:
references #44 
This comprises adding a new filter to the filter-tab of the volume-overview page as well as implementing a custom FilterSelect-component which handles the filter's behaviour, i.e. opening a modal with the openlayers-map with the option to select images. 

- [x] Add a new filter option "filter by location" to the volume overview
- [x] When the filter is selected, this opens a modal with the geo map (similar to the current map but without sidebar)
- [x] The map shows image locations which can be selected with a drawn rectangle
- [x] remove old geo-filter

mtiessen1175 commented on src/resources/assets/js/volumes/geoFilter.js:
```diff
@@ -30,4 +32,48 @@ if (Array.isArray(VolumeFilters)) {
+                };
+            },
+            created() {
+                // get all image + coordinate information from volume-images 
```
Api request

mtiessen1175 commented on src/resources/assets/js/geo/components/filterByLocationComponent.vue:
```diff
@@ -0,0 +1,54 @@
+        select(item) {
+            this.selectedItem = item;
+        },
+        gotItems(response) {
```
2. gotItems-function saves image-data as variable "items" and passes it on to geo-map-modal (3.)

mtiessen1175 commented on src/resources/assets/js/geo/components/geoMapModal.vue:
```diff
@@ -0,0 +1,107 @@
+        trigger: function() {
+            this.showModal = true;
+        },
+        // upon change in items, call getOrigin and overwrite "allImages"-variable with content of items
```
3. When "items" is updated, pass it on to GeoMap-mixin (4.)

mtiessen1175 commented on src/resources/assets/js/geo/mixins/geoMap.vue:
```diff
@@ -84,14 +84,25 @@ export default {
+        // if called from within the geoMapModal, use items-variable instead
+        if(typeof(this.items) === 'undefined') {
+            this.allImages = biigle.$require('geo.images');
+        } else {
```
4. try to pass image-data to this mixin

mzur commented on src/resources/assets/sass/main.scss:
```diff
@@ -1,4 +1,4 @@
 
 @import 'show';
-@import 'components/image-map';
+@import 'components/image-map';
```
```suggestion
```

mzur commented on src/resources/assets/js/volumes/geoFilter.js:
```diff
@@ -1,33 +1,49 @@
                     this.$emit('refresh', this.rule);
                 });
+                // trigger rule-refresh, see FilterSelect-Component of this filter
+                window.addEventListener('storageUpdate', () => {
+                    this.$emit('refresh', this.rule);
+                });
```
This may not be required (see above).

mtiessen1175 replied:
resolved 6c7af8c877c9dfe1249ab10990be473a756e65ca

mzur reviewed (changes requested):
Please remove the code to select images in the (old) geo map (including the localStorage stuff).

I don't remember what we said about updating the filter, did we say that the previous selection should persist? If not, we don't need the custom event and session store (see comments below). Instead the selected IDs can be passed on with the `select` event and we could allow multiple location filters.

mtiessen1175 commented on src/Http/Controllers/Api/FileCoordinatesController.php:
```diff
@@ -0,0 +1,45 @@
+     * @apiParam {Number} id The volume ID
+     * 
+     * @apiSuccessExample {json} Success response:
+     * {
+     *    "0": {
+     *           'id': 1, 
+     *           'lat': -7.0772818384026,
+     *           'lng': -88.464813199987
+     *         },
+     *    "1": {
+     *           'id': 2,
+     *           'lat': -7.0777,
+     *           'lng': -66.666
+     *          }
+     * }
```
Yes, since I wanted to reuse the imageMap.vue component I figured that I need to provide an images structure with id, lat and lng properties.

mtiessen1175 commented on .gitignore:
```diff
@@ -1 +1,2 @@
 node_modules
+/src/public/mix-manifest.json
```
This was a confusion I had when I started to work on this module. And I forgot to delete that line... I'll change that in the next commit.

mtiessen1175 replied:
resolved 457ee0581be87125999e8e803303a578fc6ba655

mtiessen1175 commented on src/public/assets/scripts/volumes.js.LICENSE.txt:
I didn't add it manually. My npm version is 8.5.1. Should I leave that as is or delete the file?

mzur replied:
I have 8.19. No idea why this is added. If you remove the file and then run `npm run watch` or `npm run prod`, is it added again?

mtiessen1175 replied:
Yes, `npm run prod` adds the file.

mzur replied:
Ok then let's keep it.

mtiessen1175 commented on src/resources/assets/js/geo/components/filterByLocationComponent.vue:
```diff
@@ -0,0 +1,61 @@
+<template>
+    <div class="filter-select">
+        <geo-map-modal id="gmm" :text="text" :trigger="trigger" :volumeId="volumeId" v-on:on="submit"></geo-map-modal>
```
Its not required anymore.

mtiessen1175 replied:
resolved 79072e40b224e8f671cd55bcd39c814eeeab63d2

mtiessen1175 commented on src/resources/assets/js/geo/components/filterByLocationComponent.vue:
```diff
@@ -0,0 +1,61 @@
+import GeoMapModal from './geoMapModal.vue';
+
+// create custom event to update rule upon changes in sessionStorage
+const customEvent = new Event('storageUpdate', {
+    bubbles: true,
+    cancelable: true,
+    composed: false
+  });
```
Yes, I agree, appending another "filter by location" rule feels like the most intuitive way! But as I understood it, we agreed on "replacing" the rule if there is already a "filter by location" rule in the list. This is because one cannot differentiate the individual "filter by location" rules from each other (there is no hint like labelname or username as in the other selection-filters).

mtiessen1175 replied:
Regarding the second point: I get the problem with global events like that. I did it this way because I could not really figure out how to pass the event in a "Vue way" from the filterByLocationComponent to the listComponent of the [geoFilter.js](https://github.com/biigle/geo/blob/301f08a24f472cb70218807ff8fc9d194fc33f83/src/resources/assets/js/volumes/geoFilter.js#L14C9-L14C22). This is needed, because only the listComponent of the filter is able to call the refreshRule-function (see [link](https://github.com/biigle/core/blob/169200ebe4b9b38ce2bda7e64f90c24af93e8e4e/resources/views/volumes/show/filters.blade.php#L39)). If you have a suggestion for me here, I would try to implement it differently.

mzur commented on src/resources/assets/js/volumes/geoFilter.js:
```diff
@@ -1,33 +1,49 @@
         types: ['image'],
-        label: 'geo selection',
+        label: 'filter by location',
         help: "All images that were (not) selected on the world map.",
         listComponent: {
             mixins: [FilterList],
             data() {
                 return {
-                    name: 'geo selection',
+                    name: 'filter by location',
```
I just noticed that the filter name does not match the naming scheme very well. In the active filter list, this will say "images with filter by location". Maybe call it "location selection" or even "geo selection" again? So it will say "images with geo selection".

mtiessen1175 replied:
The filter name has been changed back to "geo selection" (85e7b2c009c1c6295de5226b58aa41533ccf275c)

mtiessen1175 commented on src/resources/assets/js/volumes/geoFilter.js:
```diff
@@ -1,33 +1,49 @@
+                // only triggered when called from different session/window
                 window.addEventListener('storage', () => {
                     this.$emit('refresh', this.rule);
                 });
```
resolved 6c7af8c877c9dfe1249ab10990be473a756e65ca

mtiessen1175 commented on src/resources/assets/js/volumes/geoFilter.js:
```diff
@@ -1,33 +1,49 @@
+            mixins: [FilterSelect],
+            components: {
+            },
+            data() {
+                return {
+                    text: "<em>Hint:</em> Select image locations on the volume map by drawing an encompassing rectangle. To do this, press and hold <kbd>Ctrl</kbd> as well as the left mouse button and move the cursor on the map.",
+                };
             },
```
It is shown in the modal of the geo-filter (geoMapModal.vue)

mzur replied:
I see. Why set this as a property instead of adding it to the model template directly?

mtiessen1175 replied:
Has been changed 2c1d96e01657c8566d42cef264557909bf4a0108

mtiessen1175 replied:
Has been changed 2c1d96e01657c8566d42cef264557909bf4a0108

mtiessen1175 commented on src/resources/assets/js/geo/components/filterByLocationComponent.vue:
```diff
@@ -0,0 +1,61 @@
+<template>
+    <div class="filter-select">
+        <geo-map-modal id="gmm" :text="text" :trigger="trigger" :volumeId="volumeId" v-on:on="submit"></geo-map-modal>
```
Implemented these changes to improve code comprehension (ee5a187d8d0b5929d0e3d50fe13968d293423e5a)

mzur commented on src/resources/assets/js/geo/components/filterListComponent.vue:
You could also use the original filterListComponent as a mixin here and just override `dataName`.

However, I think I'd like to modify the base filterListComponent here, so the "x images" or "x videos" is displayed for *all* filter rules, not only the geo rules. You should be able to use `this.rule.sequence.length` to get the number of files in the filterListComponent.

mtiessen1175 replied:
> You could also use the original filterListComponent as a mixin here and just override dataName.

Thanks for pointing out! I wasn't aware of the hierarchy-behaviour of vue to overwrite mixin-functions/variables.
> However, I think I'd like to modify the base filterListComponent here...

You can find the local changes in commit fe71ba927e5c9139f8079b9d3351476a121bd2be and in core [here](https://github.com/biigle/core/pull/713#issue-2026192608)

mzur commented on src/resources/assets/js/geo/components/filterListComponent.vue:
```diff
@@ -0,0 +1,44 @@
+    methods: {
+        createName(ids) {
+            let typeForm = ids.length === 1 ? " image" : " images";
+            return "".concat("(", ids.length, typeForm, ")");
```
You can also use template literals in JS:

```suggestion
            return `(${ids.length} ${typeForm})`;
```

mtiessen1175 replied:
resolved fe71ba927e5c9139f8079b9d3351476a121bd2be

mzur commented on src/resources/assets/js/geo/components/filterByLocationComponent.vue:
```diff
@@ -33,17 +26,22 @@ export default {
+            // pass the array of selected IDs as JSON-string in order to work properly in the hasRule()-check later on.
+            // hasRule() function checks whether rule-attributes diverge between existing and new rules
+            // e.g. newRule.data === oldRule.data (does not assert equality for array- or object-comparison)
+            let ids = JSON.parse(sessionStorage.getItem(key)) || [];
+            this.selectedItem = JSON.stringify(ids);
```
Why the detour through session storage? Is there a way to return the data directly from the modal (e.g. instead of `key` here)?

If the data persists in session storage, the images will be selected again when the modal opens up the next time. I think this is counter-intuitive since we allow multiple selections now.

mzur commented on src/resources/assets/js/geo/components/filterByLocationComponent.vue:
```diff
@@ -33,17 +26,22 @@ export default {
+            // pass the array of selected IDs as JSON-string in order to work properly in the hasRule()-check later on.
+            // hasRule() function checks whether rule-attributes diverge between existing and new rules
+            // e.g. newRule.data === oldRule.data (does not assert equality for array- or object-comparison)
+            let ids = JSON.parse(sessionStorage.getItem(key)) || [];
+            this.selectedItem = JSON.stringify(ids);
```
You could also update the biigle/core code to handle the array equality check correctly.

mtiessen1175 replied:
pull request [here](https://github.com/biigle/core/pull/713#issue-2026192608)

mzur commented on src/resources/assets/js/geo/components/geoMapModal.vue:
```diff
@@ -1,27 +1,25 @@
         size="lg"
-        v-on:hide="callback"
       >
         <p v-html="text"></p>
```
Please show the text below the map and with `class="text-muted"`.

mtiessen1175 replied:
resolved 2c1d96e01657c8566d42cef264557909bf4a0108

mzur commented on src/resources/assets/js/geo/components/geoMapModal.vue:
```diff
@@ -65,39 +64,40 @@ export default {
+            },
+            (response) => {
+                return this.handleErrorResponse(response);
+            });
```
`finishLoading()` should be called in a `finally()`, otherwise loading would not finish on error:

```js
CoordApi.get({id: this.volumeId})
   .then(r => this.images = r.body, this.handleErrorResponse)
   .finally(this.finishLoading);
```

mtiessen1175 commented:
> Please still remove the code to select images in the (old) geo map (including the localStorage stuff) 😉

Sorry, I missed the last review-comment you posted. Regarding that one:

> I don't remember what we said about updating the filter, did we say that the previous selection should persist? If not, we don't need the custom event and session store (see comments below). Instead the selected IDs can be passed on with the select event and we could allow multiple location filters.

Yes we said that the previous selection should be saved in sessionStorage. But I understand that it is not necessary anymore after we decided to use multiple filters.

> Also, please leave the conversations unresolved. This makes it easier for me to go through them and resolve them one by one when I re-review your changes.

Ok :+1:

mzur commented on tests/Http/Controllers/Api/FileCoordinatesControllerTest.php:
I think I mentioned at some point that this test could randomly fail because the database does not guarantee the ordering by ID. You could test with a single image or check if an image of antother volume is *not* included in the response.

mzur reviewed (changes requested):
I see that you removed the ability to dismiss the modal with a click on the backdrop. But I can still dismiss the modal with <kbd>Esc</kbd>. When I do this, I can't add the rule by pressing on "add rule" any more.

Please adapt the [style](https://github.com/biigle/geo/blob/518ea376b3cffd6dd6e09a5ea8855554f729d125/src/resources/assets/js/geo/components/imageMap.vue#L151) of the selection box from the (old) geo map. Somehow the style is different (old is orange, modal is blue) although the same component is used.

Please remove the stuff with `localStorage` in [volumeMap](https://github.com/biigle/geo/blob/4856b3d0a73283e8fbb7c6be4a63bda40cf439b1/src/resources/assets/js/geo/volumeMap.vue) and disable `selectable` [there](https://github.com/biigle/geo/blob/4856b3d0a73283e8fbb7c6be4a63bda40cf439b1/src/resources/views/volumes/show.blade.php#L21).

All files you added in "resources/assets/js/geo/" here should be moved to "volumes" because they are only meant for the volume overview and not the (old) geo map views.

mtiessen1175 commented on src/resources/assets/js/geo/components/filterByLocationComponent.vue:
```diff
@@ -25,18 +25,17 @@ export default {
             this.hideModal();
             // pass the array of selected IDs as JSON-string in order to work properly in the hasRule()-check later on.
             // hasRule() function checks whether rule-attributes diverge between existing and new rules
             // e.g. newRule.data === oldRule.data (does not assert equality for array- or object-comparison)
```
resolved 8b73b9e80183ee4c95de2e2ab2df3506c4b2c423

mtiessen1175 commented:
> I see that you removed the ability to dismiss the modal with a click on the backdrop. But I can still dismiss the modal with Esc. When I do this, I can't add the rule by pressing on "add rule" any more.

fixed 584eb0febd5163e8d4edac99717b0ec4703d6c57

> Please adapt the [style](https://github.com/biigle/geo/blob/518ea376b3cffd6dd6e09a5ea8855554f729d125/src/resources/assets/js/geo/components/imageMap.vue#L151) of the selection box from the (old) geo map. Somehow the style is different (old is orange, modal is blue) although the same component is used.

I added the public "main.css" to the geo-filter manually (b7c31dd8c4d4fc17e46a731f915bf4f738769f66). In the (old) geo-map these styles were made available through `<link href="{{ cachebust_asset('vendor/geo/styles/main.css') }}" rel="stylesheet">` in the blade.php files. As I am not sure whether this is a valid way of doing it, I also provided a solution of injecting the styles via mixin (1ebe94f3fc7b571c961aece22db0a299fe3b72db). Maybe you could elaborate on which way you prefer and why? 

> Please remove the stuff with `localStorage` in [volumeMap](https://github.com/biigle/geo/blob/4856b3d0a73283e8fbb7c6be4a63bda40cf439b1/src/resources/assets/js/geo/volumeMap.vue) and disable selectable [there](https://github.com/biigle/geo/blob/4856b3d0a73283e8fbb7c6be4a63bda40cf439b1/src/resources/views/volumes/show.blade.php#L21).

resolved 6419f2f9ea930cc170e99d008a0a47b01d6f8a7e
> All files you added in "resources/assets/js/geo/" here should be moved to "volumes" because they are only meant for the volume overview and not the (old) geo map views.

This structure makes much more sense now :+1:  6b97869fe5060772d388fca1d5df1a1559644c7f and 50f93c1a10c778a292c8fe88c2fd0da5e2238dfd

mzur reviewed (changes requested):
Only minor things left now!

The way you handled the `main.css` is fine, too. I thought the select box is styled in JS (like the point features) but it's actually CSS.

mtiessen1175 commented on tests/Http/Controllers/Api/FileCoordinatesControllerTest.php:
```diff
@@ -38,13 +41,21 @@ public function testIndexImage()
             ->assertStatus(200);
-}
+    }
 }
```
resolved e1a91ddb951c81795eef42752a5f2d54ec483e95

mzur reviewed (approved):
Good job :+1:
